RFCs/0017-TestCase-FullyQualifiedName.md

```diff
@@ -1,49 +1,132 @@
-# 0017 Guidelines for setting the FullyQualifiedName on TestCase
+# 0017 Guidelines for setting the `FullyQualifiedName` on TestCase
 
 ## Summary
-This note clarifies the guidelines that test adapters should consider when setting the FullyQualifiedName on test cases before returning them from the discovery process.
+This note standardizes the form of `FullyQualifiedName` that test adapters should use when setting the `FullyQualifiedName` on test cases before returning them from the discovery process.
 
 ## Overview
-There are a broad variety of choices that current Test Adapter implementers have made when setting the FullyQualifiedName on TestCase objects created during test discovery. This note lays out some basic guidelines to make the usage of FullyQualifiedName more useful across adapters. These guidelines are focused on languages that support namespaces, classes and methods for structuring test cases (e.g. C# & VB). Other languages may not fit these guidelines.
+There are a broad variety of choices that current Test Adapter implementers have made when setting the `FullyQualifiedName` on TestCase objects created during test discovery. By standardizing the format of `FullyQualifiedName`, components in the vstest infrastructure can be implemented more reliably. These guidelines are focused on unit tests written in managed languages (e.g. C# & VB). Other languages may derive less benefit from adopting these guidelines.
```

Comment:
Existing RFC for FQN talks about the recommended FQN form and the corresponding normalized form. Since this PR also focuses on managed languages only, what are the additional problems this PR solves?

I can see few additional features which this PR will introduce like
1. Better extensibility for languages other than managed, 
2. Generalized FQN across adapters.

But apart from feature additions, what existing problems this PR solves which current FQN RFC is not able to? This proposed RFC causes lot of compat issues. Please check below comments for details.

Reply:
1. For managed languages, it allows us to precisely find a particular method associated with a test case. This provides a few performance benefits...
    1. We can relate tests to source code without needing any additional information outside of the FQN. This means we don't have to incur the cost of PDB processing, and it provides performance benefits in the IDE.
    1. For fast execution of a few tests (out of a large test set), we can use the FQN for execution-by-name. The standardized FQN has enough information to select the test without performing a full discovery. If we know that a method is a test method, then it even allows test execution without any discovery at all. This is helpful for Live Unit Testing as well as the IDE.

This spec would replace the need for normalization. Normalization is not a sufficient solution for the following reasons.
1. Because it is implemented using test filters, it must use RegExp processing. Regular expressions do not support recursion and so there are some common cases that cannot be properly handled, such as nested parenthetical expressions, or nested generics. In addition, string parameters cause issues. We have run into bugs in customer scenarios and we cannot fix them by only performing RegExp filtering.
2. There are a few cases, such as explicit interface implementation, where the format of the existing FQN (and normalized form as well) cannot distinguish between the type's namespace and the method's interface. This is a result of using periods to separate the segments.